File: libs/edis-ts-contracts/src/ingest.ts
/**
 * Zod mirror of `edis_contracts/ingest.py` (L1 -> L2 ingestion contracts).
 *
 * The {@link IngestEnvelopeSchema} is the stable boundary between untrusted
 * source reality and the platform. Per-domain payloads use `.strict()` to mirror
 * Pydantic's `extra="forbid"`.
 */
import { z } from "zod";
import { anyMap, datetime, schemaVersion, strMap, uuid } from "./common.js";

/** Pydantic `Domain = Literal["sales", "ops", "customer"]`. */
export const DomainSchema = z.enum(["sales", "ops", "customer"]);
export type Domain = z.infer<typeof DomainSchema>;

/** Mirror of `SalesPayloadV1` (`extra="forbid"` -> `.strict()`). */
export const SalesPayloadV1Schema = z
  .object({
    order_id: z.string(),
    customer_id: z.string(),
    sku: z.string(),
    qty: z.number().int(),
    unit_price: z.number(),
    currency: z.string().default("USD"),
    region: z.string().nullish(),
    channel: z.string().nullish(),
    order_ts: datetime(),
  })
  .strict();
export type SalesPayloadV1 = z.infer<typeof SalesPayloadV1Schema>;

/** Mirror of `OpsPayloadV1`. */
export const OpsPayloadV1Schema = z
  .object({
    service: z.string(),
    region: z.string().nullish(),
    level: z.enum(["info", "warn", "error"]).default("info"),
    status_code: z.number().int().nullish(),
    latency_ms: z.number().nullish(),
    message: z.string().nullish(),
    event_ts: datetime(),
  })
  .strict();
export type OpsPayloadV1 = z.infer<typeof OpsPayloadV1Schema>;

/** Mirror of `CustomerPayloadV1`. */
export const CustomerPayloadV1Schema = z
  .object({
    customer_id: z.string().nullish(),
    session_id: z.string(),
    event: z.string(),
    region: z.string().nullish(),
    channel: z.string().nullish(),
    props: strMap().default({}),
    event_ts: datetime(),
  })
  .strict();
export type CustomerPayloadV1 = z.infer<typeof CustomerPayloadV1Schema>;

/**
 * Mirror of `IngestEnvelope` -- the one envelope every record entering the bus is
 * wrapped in. Pydantic config is `frozen=True, extra="forbid"`; on the TS side we
 * mark the inferred type `Readonly` via `.strict()` + consumers treating it as
 * immutable.
 */
export const IngestEnvelopeSchema = z
  .object({
    event_id: uuid(),
    idempotency_key: z.string(),
    schema_ref: z.string(),
    domain: DomainSchema,
    tenant_id: z.string(),
    source_system: z.string(),
    ingest_ts: datetime(),
    event_ts: datetime(),
    producer: z.string().default("ingestion"),
    trace_context: strMap().default({}),
    is_synthetic: z.boolean().default(true),
    anomaly_label: z.string().nullish(),
    payload: anyMap(),
    schema_version: schemaVersion(),
  })
  .strict();
export type IngestEnvelope = z.infer<typeof IngestEnvelopeSchema>;

/** Mirror of `DLQRecord` (dead-letter at the ingestion edge). */
export const DLQRecordSchema = z
  .object({
    dlq_id: uuid(),
    tenant_id: z.string().nullish(),
    stage: z.enum(["ingest", "integration"]).default("ingest"),
    domain: DomainSchema.nullish(),
    source_system: z.string().nullish(),
    raw: z.unknown().nullable().default(null),
    error_type: z.string(),
    error_detail: z.string(),
    occurred_at: datetime(),
    trace_id: z.string().nullish(),
    schema_version: schemaVersion(),
  })
  .strict();
export type DLQRecord = z.infer<typeof DLQRecordSchema>;

/** Mirror of `QuarantinedRecord` (integration-layer DQ failure). */
export const QuarantinedRecordSchema = z
  .object({
    quarantine_id: uuid(),
    tenant_id: z.string(),
    stage: z.literal("integration").default("integration"),
    reason: z.string(),
    dq_failures: z.array(z.string()).default([]),
    raw: z.unknown().nullable().default(null),
    occurred_at: datetime(),
    schema_version: schemaVersion(),
  })
  .strict();
export type QuarantinedRecord = z.infer<typeof QuarantinedRecordSchema>;

File: libs/edis-contracts/edis_contracts/ingest.py
"""L1 -> L2 ingestion contracts.

The :class:`IngestEnvelope` is the stable boundary between *untrusted source
reality* and *the platform*. Per-domain payloads are validated at the edge with
``extra="forbid"`` so source schema drift is caught the moment it arrives.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

Domain = Literal["sales", "ops", "customer"]


class SalesPayloadV1(BaseModel):
    """Validated source sales record (currency/timestamps already coerced)."""

    model_config = ConfigDict(extra="forbid")

    order_id: str
    customer_id: str
    sku: str
    qty: int
    unit_price: float
    currency: str = "USD"
    region: str | None = None
    channel: str | None = None
    order_ts: datetime


class OpsPayloadV1(BaseModel):
    """Validated operations-log record (feeds error_rate / latency_p95)."""

    model_config = ConfigDict(extra="forbid")

    service: str
    region: str | None = None
    level: Literal["info", "warn", "error"] = "info"
    status_code: int | None = None
    latency_ms: float | None = None
    message: str | None = None
    event_ts: datetime


class CustomerPayloadV1(BaseModel):
    """Validated customer-activity record (feeds page_views / sessions)."""

    model_config = ConfigDict(extra="forbid")

    customer_id: str | None = None
    session_id: str
    event: str
    region: str | None = None
    channel: str | None = None
    props: dict[str, str] = Field(default_factory=dict)
    event_ts: datetime


class IngestEnvelope(BaseModel):
    """The one envelope every record entering the bus is wrapped in.

    ``idempotency_key`` is deterministic and replay-safe; ``trace_context``
    carries the W3C traceparent so a single record is traceable across all
    seven layers.
    """

    model_config = ConfigDict(frozen=True, extra="forbid")

    event_id: UUID
    idempotency_key: str
    schema_ref: str
    domain: Domain
    tenant_id: str
    source_system: str
    ingest_ts: datetime
    event_ts: datetime
    producer: str = "ingestion"
    trace_context: dict[str, str] = Field(default_factory=dict)
    is_synthetic: bool = True
    anomaly_label: str | None = None
    payload: dict[str, Any]
    schema_version: int = 1


class DLQRecord(BaseModel):
    """Dead-letter record for data that failed validation at the ingestion edge."""

    model_config = ConfigDict(extra="forbid")

    dlq_id: UUID
    tenant_id: str | None = None
    stage: Literal["ingest", "integration"] = "ingest"
    domain: Domain | None = None
    source_system: str | None = None
    raw: Any = None
    error_type: str
    error_detail: str
    occurred_at: datetime
    trace_id: str | None = None
    schema_version: int = 1


class QuarantinedRecord(BaseModel):
    """Record that parsed but failed integration-layer data-quality checks.

    Every input record terminates in exactly one of {canonical store, quarantine}
    -- never silently dropped, never double-counted.
    """

    model_config = ConfigDict(extra="forbid")

    quarantine_id: UUID
    tenant_id: str
    stage: Literal["integration"] = "integration"
    reason: str
    dq_failures: list[str] = Field(default_factory=list)
    raw: Any = None
    occurred_at: datetime
    schema_version: int = 1
